pkg/cri/sbserver/runtime_config.go

```diff
@@ -0,0 +1,83 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package sbserver
+
+import (
+	"context"
+	"sort"
+
+	"github.com/containerd/containerd/log"
+	runcoptions "github.com/containerd/containerd/runtime/v2/runc/options"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+// RuntimeConfig returns configuration information of the runtime.
+func (c *criService) RuntimeConfig(ctx context.Context, r *runtime.RuntimeConfigRequest) (*runtime.RuntimeConfigResponse, error) {
+	resp := &runtime.RuntimeConfigResponse{
+		Linux: &runtime.LinuxRuntimeConfiguration{
+			CgroupDriver: c.getCgroupDriver(ctx),
+		},
+	}
+	return resp, nil
+}
+
+func (c *criService) getCgroupDriver(ctx context.Context) runtime.CgroupDriver {
+	// Go through the runtime handlers in a predictable order, starting from the
+	// default handler, others sorted in alphabetical order
+	handlerNames := make([]string, 0, len(c.config.ContainerdConfig.Runtimes))
+	for n := range c.config.ContainerdConfig.Runtimes {
+		handlerNames = append(handlerNames, n)
+	}
+	sort.Slice(handlerNames, func(i, j int) bool {
+		if handlerNames[i] == c.config.ContainerdConfig.DefaultRuntimeName {
+			return true
+		}
+		if handlerNames[j] == c.config.ContainerdConfig.DefaultRuntimeName {
+			return false
+		}
+		return handlerNames[i] < handlerNames[j]
+	})
+
+	// Fall back to finding cgroup driver setting from other runtime handlers
+	for _, handler := range handlerNames {
+		opts, err := generateRuntimeOptions(c.config.ContainerdConfig.Runtimes[handler])
+		if err != nil {
+			log.G(ctx).Debugf("failed to parse runtime handler options for %q", handler)
+			continue
+		}
+		if d, ok := getCgroupDriverFromRuntimeHandlerOpts(opts); ok {
+			return d
+		}
+		log.G(ctx).Debugf("runtime handler %q does not provide cgroup driver information", handler)
+	}
+
+	// If no runtime handlers have a setting, return systemd as the default
```

Comment:
Needs to check whether systemd is running

Reply:
Thanks @AkihiroSuda, I think this is the right thing to do. And, now that I think of it, maybe we should just drop the `SystemdCgroup` setting from the runc handler and just rely on auto-detection everywhere. Using `cgroupfs` when `system.slice` and folks are around feels like a mis-configuration to me 🤔

Reply:
When `SystemdCgroup` is unspecified it can just detect the driver automatically.
Explicit configuration of `SystemdCgroup` should be still supported for compatibility reason.

Reply:
> Using cgroupfs when system.slice and folks are around feels like a mis-configuration to me 🤔

Still legit when the systemd unit is `Delegate`-d.

Reply:
> Still legit when the systemd unit is `Delegate`-d.

Hmm, that's true, in principle. But afaiu this scenario is not supported by kubelet at least. And, I think it's still a mis-configuration if the unit has `Delegate=yes` but you have `SystemdCgroup=true` (or vice versa if `Delegate=off` but you'd have `SystemdCgroup=false`. So, even in this scenario the right path would seem to be auto-detection (another level of auto-detection, i.e. auto-detection on the pod-sandbox unit level). WDYT @AkihiroSuda?

Reply:
I implemented systemd autodetection in case no runc plugins are configured (i.e. we cannot get the setting from there).

Regarding whether `SystemdCgroup` is explicitly set in the configuration, I don't think there's any way to detect that. It's a bool that's false by default so 🤷‍♀️